File: Recuperatorio.pl
% de los medicos sabemos su nombre , edad, especialidad
medico(juan, 25, pediatra).
medico(maria, 30, ginecologa).
medico(maria, 30, obtetricia).
medico(maria, 30, gastroenterologa).
medico(sofia, 20, enfermera).
medico(sofia, 20, anestesista).


%medico especialidad años en esa especialidad
anios(juan, pediatra, 35).
anios(maria, ginecologa, 5).
anios(maria, obstetricia, 2).
anios(sofia, enfermera, 1).

%de que se ocupa cada profesion 
ocupacion(oncologia,cancer).
ocupacion(obstetricia, embarazo).
ocupacion(obstetricia, parto).
ocupacion(gastroenterologia, higado).
ocupacion(gastroenterologia, estomago).

%paciente compuesto por nombre y sintomas 
paciente(estela, higado).
paciente(estela, embarazo).
paciente(cristian, estomago).

%------------------------------------

%Punto 0--

%De que especialidad no Hay Medicos
nohayMedicos(Especialidad):-
    not(medico(_,_,Especialidad)).

%Punto 0--

% medicos con mas de 1 especialidad
masDeUnaEspecialidad(Medico):-
    medico(Medico,_,Especialidad1),
    medico(Medico,_,Especialidad2),
    Especialidad1 \= Especialidad2.

%Punto 1--

% si un medico tiene mas años en una Especialidad que de su edad
inconsistencia(Medico):-
    medico(Medico,Edad,Especialidad),
    anios(Medico,Especialidad,Anios),
    Anios > Edad.

%Punto 2-- 

% que son los que la mayor parte de su vida han desempeñado dicha especialidad. 
muchaExperiencia(Medico):-
    medico(Medico,Edad,Especialidad),
    anios(Medico,Especialidad,Anios),
    Anios > (Edad*0.5).

%Punto 3-- 

%Buscar doctor segun el sintoma del paciente
doctorSegunPaciente(Paciente):-
    paciente(Paciente,Sintoma),
    forall(medico(Medico,_,Especialidad), ocupacion(Especialidad,Sintoma)).

%Los predicados son inversibles si al preguntar podemos saber de forma generica mediante un parametro, quienes cumplen la condicion del predicado

%Ejemplos----

%Punto 0

%masDeUnaEspecialidad(Medico).
%Medico = maria

%Punto 1

%inconsistencia(Medico).
%Medico = juan

%Punto 2

%muchaExperiencia(Medico).
%Medico = juan

File: Recuperatorio.hs
import Data.List (intersect)
-- defino el data Medico con los atributos de cada medico
data Medico = UnMedico{
    nombre :: String,
    especialidades :: [String],
    añosEspecialidad :: [Int],
    edad :: Int
} deriving(Show, Eq, Ord)

    
-- Ejemplos de algunos medicos 
maria :: Medico
maria = UnMedico {
    nombre = "Maria",
    especialidades = ["ginecologa", "obtetricia" , "gastroenterologa"],
    añosEspecialidad = [ 5 , 6 , 7] , --asumo que la lista esta en orden de acuerdo a la lista especialidad--
    edad = 30
}

juan :: Medico
juan = UnMedico {
    nombre = "Maria",
    especialidades = ["pediatra"],
    añosEspecialidad = [ 5 , 6, 7], --asumo que la lista esta en orden de acuerdo a la lista especialidad--
    edad = 25
}

-- no hay medicos de esas especialidades
noHayMedicos :: [Medico] -> [[String]] -> Bool
noHayMedicos  medicos especialidades =  intersect (especialidades) (filtrarEspecialidad medicos) ==  []

filtrarEspecialidad :: [Medico] -> [[String]]
filtrarEspecialidad medicos = map especialidades medicos

concatenarListas :: [Medico] -> [String]
concatenarListas medicos = concat (filtrarEspecialidad medicos)

-- medicos con mas de 1 especialidad
masDeUnaEspecialidad :: Medico -> Bool
masDeUnaEspecialidad medico = length (especialidades medico) >= 2 
--------------------------------------------------------------------------------
--nuevaEspecialidad (nutricionista)
--agrego a la lista del año especialidad un cero por no tener experiencia,
-- asumiendo que la lista armada coincide el orden de cada especialidad con el orden de la lista de añosEspecialidad
nutricionista :: Medico -> Medico
nutricionista medico | esNutricionista medico = medico 
                     | otherwise = medico {especialidades = especialidades medico  ++ ["nutricionista"] , añosEspecialidad = añosEspecialidad medico ++ [0]}

-- determinar si entre sus especialidades encontramos nutricionista

esNutricionista :: Medico -> Bool
esNutricionista medico = elem "nutricionista" (especialidades medico)

-- malapraxis le quita todas las especialidades a un medico
malapraxis :: Medico -> Medico
malapraxis medico = medico { especialidades = []}

cumpleaños :: Medico -> Medico
cumpleaños medico =  medico{edad = edad medico + 1 , añosEspecialidad =  sumar1 (añosEspecialidad medico)}

sumar1 medico = filter +1 (añosEspecialidad medico) 
 
